Dojo.py

```diff
@@ -0,0 +1,425 @@
+                print(my_room.room_name.upper())
+                print("-".ljust(31, '-'))
+                occupants = []
+                for occupant in my_room.occupants:
+                    occupants.append(occupant.person_name)
+                print(', '.join(occupants))
+
+        # Create a txt file if the filename argument is passed
+        if file_name is not None:
+            my_file = open("ExternalData/" + file_name + ".txt", "w")
+            for my_room in self.all_rooms:
+                my_file.write("\n")
+                my_file.write(my_room.room_name.upper())
+                my_file.write("\n" + "-".ljust(31, '-') + "\n")
+                occupants = []
+                for occupant in my_room.occupants:
+                    occupants.append(occupant.person_name)
+                my_file.write(', '.join(occupants) + "\n")
+            my_file.close()
+            # Open the file with the default application
+            self.open_file("ExternalData/" + file_name + ".txt")
+
+    def print_unallocated(self, file_name):
+        allocated_staff = []
+        allocated_fellow_office = []
+        allocated_fellow_living_space = []
+
+        for room in self.all_rooms:
+            allocated_staff += [occupant for occupant in room.occupants
+                                if isinstance(occupant, Staff)]
+            allocated_fellow_office += [occupant for occupant in room.occupants
+                                        if isinstance(occupant, Fellow) and isinstance(room, Office)]
+            allocated_fellow_living_space += [occupant for occupant in room.occupants
+                                              if isinstance(occupant, Fellow) and isinstance(room, LivingSpace)
+                                              and occupant.opt_in is True]
+        unallocated_staff = list(set([person for person in self.all_people if isinstance(person, Staff)])
+                                 - set(allocated_staff))
+        unallocated_fellow_office = list(set([person for person in self.all_people if isinstance(person, Fellow)])
+                                         - set(allocated_fellow_office))
+        unallocated_fellow_living_space = list(set([person for person in self.all_people
+                                                   if isinstance(person, Fellow) and person.opt_in is True])
+                                               - set(allocated_fellow_living_space))
+
+        print()
+        print("Unallocated staff members")
+        print("-".ljust(31, '-'))
+        print(', '.join(self.print_person_list(unallocated_staff)))
+
+        print()
+        print("Fellows without Office Space")
+        print("-".ljust(31, '-'))
+        print(', '.join(self.print_person_list(unallocated_fellow_office)))
+
+        print()
+        print("Fellows without Living Space")
+        print("-".ljust(31, '-'))
+        print(', '.join(self.print_person_list(unallocated_fellow_living_space)))
+
```

Comment:
Consider adding more user friendly feedback for when there are no allocated people instead of printing empty tables.